### ISSUE TITLE
Clarification on Resource Callbacks

### ISSUE THREAD

**Issue by namack** [JSONAPI-Resources/jsonapi-resources#652]
I'm trying to utilize JR's [resource callbacks](https://github.com/cerebris/jsonapi-resources#callbacks) but am having trouble understanding when these callbacks fire.

When attempting to modify relationship links that are returned to include authorization details (token, timestamp, etc), I plan to first use the `create_to_many_link` event to call a method that will generate the authentication credentials, however the method doesn't seem to fire.

For example:

``` ruby
class ProductResource < JSONAPI::Resource
  attributes :name, :price

  has_many :images

  after_create_to_many_link :generate_credentials

  def generate_credentials
    # credential stuff
  end
end
```

Am I invoking this incorrectly? Is there a better way to control relationship links?
(`jsonapi-resources 0.7.0`)

**Comment by utilityboy** [comment]
@namack I'm experiencing the same thing.  None of the link callbacks seem to fire.  Did you have any luck getting this sorted?

**Comment by lgebhardt** [comment]
Is this not getting called during a create or update request? If so try `replace_to_many_links` or `replace_to_one_link`. Those are the methods that are being called when the relationships are set along with other fields. If that's the case I'll see what I can do to clear that up.

**Comment by utilityboy** [comment]
@lgebhardt for myself (and I think for @namack), I'm hoping to create a custom link for a has_one relationship when I **GET** the resource.

In my case I have something like:

```
  class MachineResource < BaseResource
    attribute :name
    has_one :current_execution_plan, class_name: 'ExecutionPlan'
    has_many :execution_plans
  end

  class Machine < ActiveRecord::Base
    has_many :execution_plans

    def fetch_current_plan
      ... logic, logic, logic ...
    end
  end
```

Which renders the following on a GET to `/machines/9`

```
{
  "data": {
    "id": "9",
    "type": "machines",
    "links": {
      "self": "http://localhost:8080/machines/9"
    },
    "attributes": {
      "name": "Machine 9"
    },
    "relationships": {
      "current-execution-plan": {
        "links": {
          "self": "http://localhost:8080/machines/9/relationships/current-execution-plan",
          "related": "http://localhost:8080/machines/9/current-execution-plan"
        }
      }
    }
  }
```

I'm trying to change the following link on the relationship from:

`self: http://localhost:8080/machines/9/relationships/current-execution-plan`

to:

`self: http://localhost:8080/execution-plants/current?machine_id=9`

Such that it can be served by this controller:

```
  class ExecutionPlansController < JSONAPI::ResourceController
    prepend_before_action :swap_id_for_current, only: [:show]

    def swap_id_for_current
      return unless params[:id] == 'current'
      return unless params[:machine_id]
      expires_now

      machine = Machine.find(params[:machine_id])
      current_plan = machine.fetch_current_plan
      params[:id] = current_plan.try(:id)
    end
  end
```

Admittedly, not the most elegant, but I'd thought I could have used one of the has `*_to_one_link` callbacks to set that up.

**Comment by lgebhardt** [comment]
These callbacks are fired when resources are being updated or created. It sounds like you are looking to customize the serialization process. There isn't really an easy way to do this now, though I'm looking to expand the serialization stage to allow for customizations.

**Comment by utilityboy** [comment]
@lgebhardt thanks for the clarification.  @namack if this inline with what you were trying to achieve as well, perhaps we can close this issue.

**Comment by namack** [comment]
@utilityboy Yes, this is inline with what I was trying to do at the time. We have since shifted strategies (to not have to customize the serialization) but it would be great to have access to customizing serialization behavior in the future.

Thanks for all the hard work @lgebhardt - this is a really great gem! We recently switched over our entire existing API to JSONAPI and we couldn't have done it without JR.
